OASIS StyleGAN - S4641799

eco9898 opened the pull request:
# Task
Create a generative model of one of the OASIS brain, ADNI brain or the OAI AKOA knee data set (see Appendix for links) using either a variant of StyleGAN /StyleGAN2 or Stable Diffusion that has a “reasonably clear image”.

In this project, Stable Diffusion was experimented with, but StyleGAN was the final implementation on the OASIS brain dataset. Stable Diffusion wasn't used due to a lack of familiarity with the model type.
# Project Files
- `dataset.py`: Data loader for the OASIS brain dataset, transforms the data for use with the model.
- `modules.py`: The StyleGAN networks as well as the layers and blocks they use.
- `predict.py`: Generates a batch of fake images that can be compared to the original dataset.
- `train.py`: Training process for the StyleGAN networks, will also generate example images and graph the loss function.
- `utils.py`: Functions used by the other scripts and global variables.
# Results
By epoch 150 I consider the model a success, however, it was starting to generate repeated and similar images. Training the model any longer would most likely cause the model to become overtrained and collapse the training simulation.
![1697679243 0313046_predicted](https://github.com/shakes76/PatternAnalysis-2023/assets/15917476/b18babda-beac-43d3-b628-f07d8b6829b1)

eco9898 committed:
Initial UNet implementation

UNet implementation based off https://github.com/aju22/DDPM/blob/main/Denoising_Diffusion_Probabilistic_Model_(DDPMs).ipynb

eco9898 committed:
Initial training function

Basic function to train the model and save image output every 10 epochs.
The model is only saved if it out performs the last best model saved.

eco9898 committed:
Skeleton Files

eco9898 committed:
Basic dataset loader

Initial dataset loader using the basic ImageFolder class

eco9898 committed:
Checkpoint saving and loading

The model now saves checkpoints along with the best model to resume training later.
Also upgraded the progress bar with tqdm support

eco9898 committed:
First attempt at stylegan2

Implementing stylegan2 modules based on https://blog.paperspace.com/implementation-stylegan2-from-scratch/#noise-mapping-network

eco9898 committed:
Modified dataset loader to include seg images

Modified dataset loader to include seg images

eco9898 committed:
Increase IO resolution to 512 from 128

Increased input and output resolution to 512 from 128
Also experimented with increasing internal resolution, but this overflowed VRAM past 40GB
side note: decreased training epochs from 300 to 30 as most models are trained by 15

eco9898 committed:
Reduce IO to 256

Reduced IO down to 256 as 40gb ram wasn't enough when validating and saving the 512 model

eco9898 committed:
Reworked loading and saving checkpoints to save latest and best data

Reworked loading and saving checkpoints to save latest and best data

eco9898 committed:
Added manually stopping training and saving progress

eco9898 committed:
Updated variables relating to IMAGE_SIZE to auto scale

Updated variables relating to IMAGE_SIZE to auto scale
Also updated generate_images to be able to generate a singular image without saving it to disk

eco9898 committed:
Created predict.py to display or compare 32 model generated images

Created predict.py to display or compare 32 model generated images.
This can be toggled to choose the latest or best model as well as to compare to validation data or display by itself

eco9898 committed:
varius bug fixes for modules, and utils

Fixed circular import between modules and utils
Fixed type error when calculating log_resolution

eco9898 committed:
Updated saving on close to not override older latest models

Closing will only save over the latest models when the current model has past the previous model. This fixes accidently loading the wrong model and deleting an older model when closing it

eco9898 committed:
Removed stable diffusion code

Removed code artifacts from stable diffusion testing and implementation

eco9898 committed:
Created README and uploaded assets

Created README and uploaded assets

eco9898 committed:
Added short description to each file

WeiDaiwow commented:
# Document
 
- Title: 
	Good
- Address problem:
	Problem mentioned, but no reasons were provided for why Style GAN is good for this task.
- Working Principle:
	Good
- Figure visualization
	Good
- dependency
	Good
- inputs and outputs description:
	input visualization not provided.
	output visualization provided.
- training plot:
	loss plot only contains single loss. Is this discriminator or generator? Should plot both
- readme using markdown:
	Good
- preprocessing description:
	Not mentioned
- Description of usage:
	Good

Suggestions:
1. the link in noise injection is broken 


# Check the files:
- modules.py
	no comments provided
- dataset.py
	not enough comments
- train.py
	not enough comments
- predict.py
	Good



# Overall:

- solve the problem:
	Yes
- TF/torch implementation:
	Yes
- Design quality:
	Good
- code comments:
	not enough comments for modules.py, dataset.py and train.py
- Difficulty: 
	styleganv2 -hard

eco9898 committed:
Fix webp file not displaying on github

eco9898 committed:
Label loss graph

eco9898 committed:
Generated noise input examples

Added noise generated example to the readme and a function to generate a noise example

eco9898 committed:
Added comments to python files

Added comments to python files as requested

eco9898 committed:
Added readme updates

Included justification for why StyleGAN was used, as well as the modifications done to pre-process images for the model, and some minor style and formatting changes.

eco9898 commented:
New commits were added to address:
1. Noise injection image not appearing
2. Adding noise input examples to the readme
3. Improving and adding comments to the python files
4. Updating the Readme content to include justification for why StyleGAN was chosen
5. Updating the Readme content to include preprocessing done on training images 
6. Updating the readme content to fix minor formatting and style issues

If any more changes are required please let me know, but I believe this addresses all the issues raised in the feedback. @WeiDaiwow

shakes76 commented:
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting minimal -1
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable/good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in InCorrect Branch) -2
Feedback required, fix branch to correct branch
Request Description good

shakes76 commented:
Cannot merge because of gitignore file and incorrect name of folder, not stable diffusion but Style GAN, please amend. Does not affect grade only the merge.

shakes76 commented:
Feedback marks possible +2 if the requested changes are made (see above).

wangzhaomxy commented:
No feedback attempt and no feedback marks granted.

eco9898 committed:
Rename folder to stylegan from stable diffusion

eco9898 commented:
Sorry, I didn't receive a notification about these comments and as semester had ended was no longer actively checking for comments. I will work on these changes now. @shakes76 and @wangzhaomxy

eco9898 committed:
Fixed header block comments

Fixed header block comments for each function and amended overall file comment block

eco9898 committed:
Merge branch 'topic-recognition' into topic-recognition

eco9898 commented:
Addressed issues relating to the pull request's branch

eco9898 committed:
Update dataset.py header block

eco9898 committed:
Update modules.py header block

eco9898 committed:
Update predict.py header block

eco9898 committed:
Update train.py header block

eco9898 committed:
Update utils.py header block

eco9898 commented:
Feedback complete and an issue was raised on edstem: https://edstem.org/au/courses/13207/discussion/1726226, which has now been answered

wangzhaomxy commented:
Feedback attempt and feedback marks granted +2.
